src/main/kotlin/org/koitharu/kotatsu/parsers/site/Manhwa18Parser.kt

```diff
@@ -0,0 +1,180 @@
+	// 2 days ago
+	// 2 weeks ago
+	// 4 years ago
+	private fun parseUploadDate(timeStr: String?): Long {
+		timeStr ?: return 0
+
+		val timeWords = timeStr.split(' ')
+		if (timeWords.size != 3) return 0
+		val timeWord = timeWords[1]
+		val timeAmount = timeWords[0].toIntOrNull() ?: return 0
+		val timeUnit = when (timeWord) {
+			"minute", "minutes" -> Calendar.MINUTE
+			"hour", "hours" -> Calendar.HOUR
+			"day", "days" -> Calendar.DAY_OF_YEAR
+			"month", "months" -> Calendar.MONTH
+			"year", "years" -> Calendar.YEAR
+			else -> return 0
+		}
+		val cal = Calendar.getInstance()
+		cal.add(timeUnit, timeAmount)
+		return cal.time.time
+	}
+
+	override suspend fun getListPage(
+		page: Int,
+		query: String?,
+		tags: Set<MangaTag>?,
+		sortOrder: SortOrder,
+	): List<Manga> {
+		val sortQuery = when(sortOrder) {
+			SortOrder.ALPHABETICAL -> "name"
+			SortOrder.POPULARITY -> "views"
+			SortOrder.UPDATED -> "last_update"
+			else -> ""
+		}
+
+		val sortType = if (sortOrder == SortOrder.ALPHABETICAL) "ASC" else "DESC"
+		val tagQuery = tags.orEmpty().joinToString(",") { it.key }
+		val url = buildString {
+			append("https://")
+			append(getDomain())
+			append("/manga-list.html?listType=pagination&page=")
+			append(page)
+			append("&artist=&author=&group=&m_status=&name=")
+			append(query.orEmpty())
+			append("&genre=$tagQuery")
+			append("&ungenre=")
+			append("&sort=")
+			append(sortQuery)
+			append("&sort_type=")
+			append(sortType)
+		}
+
+		val docs = context.httpGet(url).parseHtml()
+		val actualPage = docs.selectFirst("ul.pagination  a.active")?.text()?.toIntOrNull()
+		if (actualPage != page) {
+			return emptyList()
+		}
+
+		return docs.select(".card-body .thumb-item-flow")
+			.map {
+				val titleElement = it.selectFirstOrThrow(".thumb_attr.series-title > a")
+				val absUrl = titleElement.attrAsAbsoluteUrl("href")
+				Manga(
+					id = generateUid(absUrl.toRelativeUrl(getDomain())),
+					title = titleElement.text(),
+					altTitle = null,
+					url = absUrl.toRelativeUrl(getDomain()),
+					publicUrl = absUrl,
+					rating = RATING_UNKNOWN,
+					isNsfw = true,
+					coverUrl = it.selectFirst("div.img-in-ratio")?.attrAsAbsoluteUrl("data-bg").orEmpty(),
+					tags = emptySet(),
+					state = null,
+					author = null,
+					largeCoverUrl = null,
+					description = null,
+					source = MangaSource.MANHWA18,
+				)
+			}
+	}
+
+	override suspend fun getPages(chapter: MangaChapter): List<MangaPage> {
+		val chapterUrl = chapter.url.toAbsoluteUrl(getDomain())
+		return context.httpGet(chapterUrl).parseHtml()
+			.select("div.chapter-content > img").map {
+				val url = it.attrAsRelativeUrlOrNull("src").orEmpty()
+				MangaPage(
+					id = generateUid(url),
+					url = url,
+					referer = chapterUrl,
+					preview = null,
+					source = MangaSource.MANHWA18,
+				)
+			}
+	}
+
+	override suspend fun getTags(): Set<MangaTag> {
+		return context.httpGet("https://${getDomain()}/").parseHtml().selectFirst(".genres-menu")
```

Comment:
Maybe it should be better to throw an exception instead of returning an empty set?
```suggestion
		return context.httpGet("https://${getDomain()}/").parseHtml().selectFirstOrThrow(".genres-menu")
```